### PR TITLE
[shardformer] fix gathering output when using tensor parallelism

### PR DESCRIPTION
**Pull request by flybird11111**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

## 💥 Checklist before requesting a review

- [ ] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [ ] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/shardformer/policies/gpt2.py [hunk 1]
```diff
@@ -38,6 +38,13 @@ def preprocess(self):
             if vocab_size % world_size != 0:
                 new_vocab_size = vocab_size + world_size - vocab_size % world_size
                 self.model.resize_token_embeddings(new_vocab_size)
+        elif self.shard_config.pipeline_stage_manager is not None:
```
Comment (ver217) [inline colossalai/shardformer/policies/gpt2.py:41]: I think this is not related to pipeline parallelism, right? As it actually affects kernels rather than p2p communication. Shall we make it more general?

File: colossalai/booster/plugin/hybrid_parallel_plugin.py [hunk 2]
```diff
@@ -963,6 +964,7 @@ def __init__(
         pp_style: str = "1f1b",
         num_model_chunks: int = 1,
         enable_metadata_cache: bool = True,
+        make_vocab_size_divisible_by: int = 128,
```
Comment (ver217) [inline colossalai/booster/plugin/hybrid_parallel_plugin.py:967]: Please also update docstr

**Review by ver217 — CHANGES_REQUESTED** [review]

File: colossalai/booster/plugin/hybrid_parallel_plugin.py (hunk 2 above)
Comment (flybird11111) [reply colossalai/booster/plugin/hybrid_parallel_plugin.py:969]: updated.

Comment (flybird11111) [reply colossalai/booster/plugin/hybrid_parallel_plugin.py:969]: updated.

File: colossalai/shardformer/policies/gpt2.py (hunk 1 above)
Comment (flybird11111) [reply colossalai/shardformer/policies/gpt2.py]: updated.

File: colossalai/shardformer/policies/gpt2.py [hunk 3]
```diff
@@ -38,6 +38,13 @@ def preprocess(self):
             if vocab_size % world_size != 0:
                 new_vocab_size = vocab_size + world_size - vocab_size % world_size
                 self.model.resize_token_embeddings(new_vocab_size)
+        else:
+            # Make vocab_size divisible by `make_vocab_size_divisible_by` to select a faster CUDA kernel operator.
+            new_vocab_size = vocab_size
+            multiple = self.shard_config.make_vocab_size_divisible_by
+            while (new_vocab_size % multiple) != 0:
```
Comment (ver217) [inline colossalai/shardformer/policies/gpt2.py:45]: This is too slow

**Review by ver217 — CHANGES_REQUESTED** [review]

File: colossalai/shardformer/policies/gpt2.py (hunk 3 above)
Comment (flybird11111) [reply colossalai/shardformer/policies/gpt2.py]: updated.

File: colossalai/shardformer/policies/gpt2.py [hunk 4]
```diff
@@ -38,6 +38,11 @@ def preprocess(self):
             if vocab_size % world_size != 0:
                 new_vocab_size = vocab_size + world_size - vocab_size % world_size
                 self.model.resize_token_embeddings(new_vocab_size)
+        else:
+            # Make vocab_size divisible by `make_vocab_size_divisible_by` to select a faster CUDA kernel operator.
+            multiple = self.shard_config.make_vocab_size_divisible_by
+            new_vocab_size = (vocab_size // multiple + 1) * multiple
```
Comment (ver217) [inline colossalai/shardformer/policies/gpt2.py:44]: This is not correct when vocab_size % multiple == 0

**Review by ver217 — CHANGES_REQUESTED** [review]

**Review by ver217 — APPROVED** [review]
